File: WoofAndroid/design-system/src/main/kotlin/com/perrystreet/woof/designsystem/atomic/molecules/button/MolIconButton.kt
package com.perrystreet.woof.designsystem.atomic.molecules.button

import androidx.compose.animation.core.animateFloatAsState
import androidx.compose.foundation.background
import androidx.compose.foundation.interaction.MutableInteractionSource
import androidx.compose.foundation.interaction.collectIsPressedAsState
import androidx.compose.foundation.shape.CircleShape
import androidx.compose.material3.IconButton
import androidx.compose.runtime.Composable
import androidx.compose.runtime.getValue
import androidx.compose.runtime.remember
import androidx.compose.ui.Modifier
import androidx.compose.ui.draw.alpha
import androidx.compose.ui.draw.scale
import androidx.compose.ui.res.stringResource
import com.perrystreet.woof.designsystem.atomic._tokens.sizing.SizingRoles
import com.perrystreet.woof.designsystem.atomic.atoms.icon.AtomIcon
import com.perrystreet.woof.designsystem.atomic.atoms.icon.roles.IconColorRole
import com.perrystreet.woof.designsystem.atomic.molecules.button.roles.IconButtonRole
import com.perrystreet.woof.designsystem.theme.Theme

@Composable
fun MolIconButton(
    role: IconButtonRole,
    onTap: () -> Unit,
    modifier: Modifier = Modifier,
    isActive: Boolean = false,
    isOnScrim: Boolean = false,
    hasBackground: Boolean = false,
) {
    val interactionSource = remember { MutableInteractionSource() }
    val isPressed by interactionSource.collectIsPressedAsState()
    val alpha by animateFloatAsState(
        targetValue = when (isPressed) {
            true -> Theme.alpha.pressed
            false -> Theme.alpha.enabled
        },
        label = "icon_button_alpha",
    )
    val scale by animateFloatAsState(
        targetValue = when (isPressed) {
            true -> PressedScale
            false -> 1f
        },
        label = "icon_button_scale",
    )
    val backgroundModifier = when (hasBackground) {
        true -> Modifier.background(Theme.colors.scrimContainer, CircleShape)
        false -> Modifier
    }
    val colorRole = when {
        isActive -> role.activeColorRole
        isOnScrim -> IconColorRole.OnScrim
        else -> IconColorRole.OnSurface
    }
    val iconRes = when (isActive) {
        true -> role.activeIconRes
        false -> role.iconRes
    }
    val contentDescriptionRes = when (isActive) {
        true -> role.activeContentDescriptionRes
        false -> role.contentDescriptionRes
    }

    IconButton(
        onClick = onTap,
        interactionSource = interactionSource,
        modifier = modifier
            .then(backgroundModifier)
            .scale(scale)
            .alpha(alpha),
    ) {
        AtomIcon(
            iconRes = iconRes,
            iconSize = SizingRoles.Icon.M,
            contentDescription = stringResource(contentDescriptionRes),
            colorRole = colorRole,
        )
    }
}

private const val PressedScale = 0.85f

File: WoofIOS/SwiftPackages/DesignSystem/Sources/DesignSystem/Atomic/Molecules/Button/MolIconButton.swift
import Resources
import SwiftUI

public struct MolIconButton: View {
    @Environment(\.theme) private var theme

    private let role: IconButtonRole
    private let onTap: () -> Void
    private let isActive: Bool
    private let isOnScrim: Bool
    private let hasBackground: Bool

    private static let pressedScale: CGFloat = 0.85

    public init(
        role: IconButtonRole,
        onTap: @escaping () -> Void,
        isActive: Bool = false,
        isOnScrim: Bool = false,
        hasBackground: Bool = false
    ) {
        self.role = role
        self.onTap = onTap
        self.isActive = isActive
        self.isOnScrim = isOnScrim
        self.hasBackground = hasBackground
    }

    public var body: some View {
        Button(action: onTap) {
            AtomIcon(
                icon: isActive ? role.activeIcon : role.icon,
                iconSize: .m,
                contentDescription: isActive ? role.activeContentDescription : role.contentDescription,
                colorRole: colorRole
            )
            .frame(width: theme.sizing.interactionHeightDefault, height: theme.sizing.interactionHeightDefault)
            .background(backgroundColor, in: Circle())
        }
        .buttonStyle(MolIconButtonStyle(pressedScale: Self.pressedScale, pressedAlpha: theme.alpha.pressed, enabledAlpha: theme.alpha.enabled))
    }

    private var colorRole: IconColorRole {
        switch (isActive, isOnScrim) {
        case (true, _): role.activeColorRole
        case (false, true): .onScrim
        case (false, false): .onSurface
        }
    }

    private var backgroundColor: Color {
        switch hasBackground {
        case true: theme.colors.scrimContainer
        case false: Color.clear
        }
    }
}

private struct MolIconButtonStyle: ButtonStyle {
    let pressedScale: CGFloat
    let pressedAlpha: Double
    let enabledAlpha: Double

    func makeBody(configuration: Configuration) -> some View {
        configuration.label
            .scaleEffect(configuration.isPressed ? pressedScale : 1)
            .opacity(configuration.isPressed ? pressedAlpha : enabledAlpha)
            .animation(.default, value: configuration.isPressed)
    }
}
